homeassistant/components/ws66i/media_player.py

```diff
@@ -0,0 +1,213 @@
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the WS66i 6-zone amplifier platform from a config entry."""
+    ws66i_data: Ws66iData = hass.data[DOMAIN][config_entry.entry_id]
+
+    # Build and add the entities from the data class
+    async_add_entities(
+        Ws66iZone(
+            device=ws66i_data.device,
+            ws66i_data=ws66i_data,
+            entry_id=config_entry.entry_id,
+            zone_id=zone_id,
+            data_idx=idx,
+            coordinator=ws66i_data.coordinator,
+        )
+        for idx, zone_id in enumerate(ws66i_data.zones)
+    )
+
+    # Set up services
+    platform = async_get_current_platform()
+
+    platform.async_register_entity_service(
+        SERVICE_SNAPSHOT,
+        {},
+        "snapshot",
+    )
+
+    platform.async_register_entity_service(
+        SERVICE_RESTORE,
+        {},
+        "async_restore",
+    )
+
+
+class Ws66iZone(CoordinatorEntity, MediaPlayerEntity):
+    """Representation of a WS66i amplifier zone."""
+
+    def __init__(
+        self,
+        device: WS66i,
+        ws66i_data: Ws66iData,
+        entry_id: str,
+        zone_id: int,
+        data_idx: int,
+        coordinator: Ws66iDataUpdateCoordinator,
+    ) -> None:
+        """Initialize a zone entity."""
+        super().__init__(coordinator)
+        self._ws66i: WS66i = device
+        self._ws66i_data: Ws66iData = ws66i_data
+        self._zone_id: int = zone_id
+        self._zone_id_idx: int = data_idx
+        self._coordinator = coordinator
+        self._snapshot: ZoneStatus = None
+        self._status: ZoneStatus = coordinator.data[data_idx]
+        self._attr_source_list = ws66i_data.sources.name_list
+        self._attr_unique_id = f"{entry_id}_{self._zone_id}"
+        self._attr_name = f"Zone {self._zone_id}"
+        self._attr_supported_features = SUPPORT_WS66I
+        self._attr_device_info = DeviceInfo(
+            identifiers={(DOMAIN, str(self.unique_id))},
+            name=self.name,
+            manufacturer="Soundavo",
+            model="WS66i 6-Zone Amplifier",
+        )
+        self._set_attrs_from_status()
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle updated data from the coordinator."""
+        # This will be called for each of the entities after the coordinator
+        # finishes executing _async_update_data()
+
+        # Save a reference to the zone status that this entity represents
+        self._status = self.coordinator.data[self._zone_id_idx]
+        self._set_attrs_from_status()
+
+        # Parent will notify HA of the update
+        super()._handle_coordinator_update()
+
+    @callback
+    def _set_attrs_from_status(self) -> None:
+        status = self._status
+        sources = self._ws66i_data.sources.id_name
+        self._attr_state = STATE_ON if status.power else STATE_OFF
+        self._attr_volume_level = status.volume / float(MAX_VOL)
```

Comment:
Maybe round the result?

Reply:
I don't think so. Volume level should be between 0 and 1.